symphony-bdk-core/src/main/java/com/symphony/bdk/core/extension/ExtensionService.java

```diff
@@ -0,0 +1,135 @@
+import com.symphony.bdk.extension.BdkExtensionServiceProvider;
+
+import lombok.extern.slf4j.Slf4j;
+import org.apiguardian.api.API;
+
+import java.lang.reflect.InvocationTargetException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+/**
+ * Service class for managing extensions.
+ *
+ * @see BdkExtension
+ */
+@Slf4j
+@API(status = API.Status.EXPERIMENTAL)
+public class ExtensionService {
+
+  private final Map<Class<? extends BdkExtension>, BdkExtension> extensions;
+
+  private final ApiClientFactory apiClientFactory;
+  private final AuthSession botSession;
+  private final RetryWithRecoveryBuilder<?> retryBuilder;
+  private final BdkConfig config;
+
+  public ExtensionService(
+      @Nonnull ApiClientFactory apiClientFactory,
+      @Nullable AuthSession botSession,
+      @Nonnull RetryWithRecoveryBuilder<?> retryBuilder,
+      @Nonnull BdkConfig config
+  ) {
+    this.apiClientFactory = apiClientFactory;
+    this.botSession = botSession;
+    this.retryBuilder = retryBuilder;
+    this.config = config;
+    this.extensions = Collections.synchronizedMap(new HashMap<>());
+  }
+
+  public void register(BdkExtension extension) {
+
+    final Class<? extends BdkExtension> extClz = extension.getClass();
+
+    if (extension instanceof BdkAuthenticationAware) {
+      if (this.botSession == null) {
+        log.info("Extension <{}> uses authentication, but it has not been configured in BDK config", extClz);
+      } else {
+        log.debug("Extension <{}> uses authentication", extClz);
+        ((BdkAuthenticationAware) extension).setAuthSession(this.botSession);
+      }
+    }
+
+    if (extension instanceof BdkApiClientFactoryAware) {
+      log.debug("Extension <{}> uses the ApiClientFactory", extClz);
+      ((BdkApiClientFactoryAware) extension).setApiClientFactory(this.apiClientFactory);
+    }
+
+    if (extension instanceof BdkRetryBuilderAware) {
+      log.debug("Extension <{}> uses the RetryBuilder", extClz);
+      ((BdkRetryBuilderAware) extension).setRetryBuilder(this.retryBuilder);
+    }
+
+    if (extension instanceof BdkConfigAware) {
+      log.debug("Extension <{}> uses the configuration", extClz);
+      ((BdkConfigAware) extension).setConfiguration(this.config);
+    }
+
+    this.extensions.put(extClz, extension);
+  }
+
+  /**
+   * Registers and instantiates an extension.
+   *
+   * @param extClz Type of the extension.
+   * @throws IllegalStateException if the extension has already been registered
+   * @throws BdkExtensionException if the extension cannot be instantiated
+   * @see BdkExtension
+   */
+  public void register(Class<? extends BdkExtension> extClz) {
+    log.debug("Registering extension <{}>", extClz);
+
+    if (this.extensions.get(extClz) != null) {
+      throw new IllegalStateException("Extension <" + extClz + "> has already been registered");
```

Comment:
Should we do that kind of check in `public void register(BdkExtension extension)` as well?